platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/gossip/shadowgraph/SyncUtils.java

```diff
@@ -16,34 +16,407 @@
+                    }
+                }
+            } finally {
+                // in case an exception gets thrown, unblock the writer thread
+                eventReadingDone.countDown();
+            }
+        };
+    }
+
+    /**
+     * Checks if the time spent sending and receiving events exceeds maximum allowed. If it has, it throws an
+     * exception.
+     *
+     * @param startTime the time at which phase 3 started
+     * @throws SyncTimeoutException thrown if the time is exceeded
+     */
+    private static void checkEventExchangeTime(final long startTime) throws SyncTimeoutException {
+        final long phase3Nanos = System.nanoTime() - startTime;
+        if (phase3Nanos > SEND_AND_RECEIVE_EVENTS_MAX_TIME.toNanos()) {
+            throw new SyncTimeoutException(Duration.ofNanos(phase3Nanos), SEND_AND_RECEIVE_EVENTS_MAX_TIME);
+        }
+    }
+
+    /**
+     * Get the latest self event in the shadowgraph.
+     *
+     * @param shadowGraph the shadow graph
+     * @param selfId      the id of the node
+     * @return the latest self event in the shadowgraph, or null if there is none
+     */
+    @Nullable
+    private static ShadowEvent getLatestSelfEventInShadowgraph(
+            @NonNull final ShadowGraph shadowGraph, @NonNull final NodeId selfId) {
+
+        final List<ShadowEvent> tips = shadowGraph.getTips();
+        for (final ShadowEvent tip : tips) {
+            if (tip.getEvent().getCreatorId().equals(selfId)) {
+                return tip;
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Given a list of events we think the other node may not have, reduce that list to events that we think they do not
+     * have and that are unlikely to end up being duplicate events.
+     *
+     * <p>
+     * General principals:
+     * <ul>
+     * <li>Always send self events right away.</li>
+     * <li>Don't send non-ancestors of self events unless we've known about that event for a long time.</li>
+     * <li>Don't send ancestors of self events right away, but send them sooner than non-ancestors.</li>
+     * </ul>
+     *
+     * @param shadowGraph          the shadow graph
+     * @param selfId               the id of this node
+     * @param ancestorThreshold    the amount of time an event must be known about before we send its ancestors
+     * @param nonAncestorThreshold the amount of time an event must be known about before we send it
+     * @param now                  the current time
+     * @param eventsTheyNeed       the list of events we think they need
+     */
+    @NonNull
+    public static List<EventImpl> filterLikelyDuplicates(
+            @NonNull final ShadowGraph shadowGraph,
+            @NonNull final NodeId selfId,
+            @NonNull final Duration ancestorThreshold,
+            @NonNull final Duration nonAncestorThreshold,
+            @NonNull final Instant now,
+            @NonNull final List<EventImpl> eventsTheyNeed) {
```

Comment:
Note to reviewers: these are the only new sync functions. All other functions are basically the same as before, but renamed and moved to this file where we are keeping all of the static methods for the sync algorithm.